### ISSUE TITLE
functions:shell not emulating functions

### ISSUE THREAD

**Issue by prgwar** [firebase/firebase-tools#1456]
node: 8.13.0
firebase-tools:7.0.2
firebase-admin: 8.2.0

when I enter firebase functions:shell following output is coming 

functions: Emulator started at http://localhost:5001
after that its struck 
I have set the GOOGLE_APPLICATION_CREDENTIALS

**Comment by samtstern** [comment]
@prgwar we need a lot more information to debug this.

- Can you run the command with `--debug` and see if there is any output?
- What is the simplest functions source code that reproduces this issue so we can try it>

**Comment by sperochon** [comment]
I've got the same problem. The command 'firebase functions:shell' was working fine for months. Then I upgraded my node.js from 8.11 to 8.16, firebase tools and CLI (and lot of packages) and now it gets stucks = nothing happen after displaying
'✔  functions: Emulator started at http://localhost:5001'

Here is the log with --debug
`[2019-07-03T18:55:49.105Z] ----------------------------------------------------------------------
[2019-07-03T18:55:49.109Z] Command:       /Users/seb/.nvm/versions/node/v8.16.0/bin/node /Users/seb/.nvm/versions/node/v8.16.0/bin/firebase functions:shell --debug
[2019-07-03T18:55:49.109Z] CLI Version:   7.0.2
[2019-07-03T18:55:49.109Z] Platform:      darwin
[2019-07-03T18:55:49.109Z] Node Version:  v8.16.0
[2019-07-03T18:55:49.109Z] Time:          Wed Jul 03 2019 20:55:49 GMT+0200 (CEST)
[2019-07-03T18:55:49.109Z] ----------------------------------------------------------------------

[2019-07-03T18:55:49.119Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2019-07-03T18:55:49.119Z] > authorizing via signed-in user
[2019-07-03T18:55:49.119Z] [iam] checking project xxxxxxx-dev-eu-west3 for permissions ["firebase.projects.get"]
[2019-07-03T18:55:49.121Z] >>> HTTP REQUEST POST https://cloudresourcemanager.googleapis.com/v1/projects/xxxxxxx-dev-eu-west3:testIamPermissions  
 permissions=[firebase.projects.get]
[2019-07-03T18:55:49.967Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=UTF-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Wed, 03 Jul 2019 18:55:49 GMT, server=ESF, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, server-timing=gfet4t7; dur=663, alt-svc=quic=":443"; ma=2592000; v="46,43,39", accept-ranges=none, transfer-encoding=chunked
[2019-07-03T18:55:49.979Z] [functions] Using node@8 from host.
[2019-07-03T18:55:49.981Z] >>> HTTP REQUEST GET https://cloudresourcemanager.googleapis.com/v1/projects/xxxxxxx-dev-eu-west3  
 
[2019-07-03T18:55:50.740Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=UTF-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Wed, 03 Jul 2019 18:55:50 GMT, server=ESF, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, server-timing=gfet4t7; dur=662, alt-svc=quic=":443"; ma=2592000; v="46,43,39", accept-ranges=none, transfer-encoding=chunked
[2019-07-03T18:55:50.741Z] >>> HTTP REQUEST GET https://mobilesdk-pa.googleapis.com/v1/projects/595396935831:getServerAppConfig  
 
[2019-07-03T18:55:50.998Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=UTF-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Wed, 03 Jul 2019 18:55:50 GMT, server=ESF, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, alt-svc=quic=":443"; ma=2592000; v="46,43,39", accept-ranges=none, transfer-encoding=chunked
✔  functions: Emulator started at http://localhost:5001
[2019-07-03T18:55:51.004Z] [functions] Watching "/Users/seb/dev/git/xxxxxxx/xxxxxxx_firebase/Firebase/functions" for Cloud Functions...
[2019-07-03T18:55:51.264Z] DEBUG: Functions runtime initialized.
[2019-07-03T18:55:51.265Z] DEBUG: Disabled runtime features: {"functions_config_helper":true,"network_filtering":true,"timeout":true,"memory_limiting":true,"protect_env":true,"admin_stubs":true}
[2019-07-03T18:55:51.332Z] DEBUG: firebase-admin has been stubbed.
`

**Comment by samtstern** [comment]
@sperochon thanks for that, that's interesting.

In particular these lines are concerning to me:
```
[2019-07-03T18:55:51.265Z] DEBUG: Disabled runtime features: {"functions_config_helper":true,"network_filtering":true,"timeout":true,"memory_limiting":true,"protect_env":true,"admin_stubs":true}
[2019-07-03T18:55:51.332Z] DEBUG: firebase-admin has been stubbed.
```

If `admin_stubs` is disabled then we should not see `firebase-admin has been stubbed`

**Comment by samtstern** [comment]
@abeisgoat why don't we early-return from the entire `InitializeFirebaseAdminStubs` function if the feature is disabled?  Seems like we only check inside the firestore settings block.

**Comment by samtstern** [comment]
@sperochon does this same issue happen for you if you use the Firebase CLI version `6.12.0`:

```
npm install -g firebase-tools@6.12.0
firebase serve
```

**Comment by sperochon** [comment]
Just to let you know, I tried to isolate the problem and remove some files and finally got it working when I have only a few files. The weird thing is that it is not concerning a specific file... and that changing the region from 'europe-west1' to 'us-east1' for 1 function makes it working ?!

```
exports.evtOnCreate = functions
  .region('us-east1')
  .firestore
  .document('/evts/{evtId}')
  .onCreate(async (snap, context) => {...
```

Here is the log when it is working fine (it's exactly the same until 'DEBUG: firebase-admin has been stubbed.')
```
[debug] [2019-07-03T18:37:36.758Z] ----------------------------------------------------------------------
[debug] [2019-07-03T18:37:36.760Z] Command:       /Users/seb/.nvm/versions/node/v8.16.0/bin/node /Users/seb/.nvm/versions/node/v8.16.0/bin/firebase functions:shell
[debug] [2019-07-03T18:37:36.760Z] CLI Version:   7.0.2
[debug] [2019-07-03T18:37:36.760Z] Platform:      darwin
[debug] [2019-07-03T18:37:36.760Z] Node Version:  v8.16.0
[debug] [2019-07-03T18:37:36.761Z] Time:          Wed Jul 03 2019 20:37:36 GMT+0200 (CEST)
[debug] [2019-07-03T18:37:36.761Z] ----------------------------------------------------------------------
[debug] 
[debug] [2019-07-03T18:37:36.769Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[debug] [2019-07-03T18:37:36.769Z] > authorizing via signed-in user
[debug] [2019-07-03T18:37:36.770Z] [iam] checking project xxxxxxx-dev-eu-west3 for permissions ["firebase.projects.get"]
[debug] [2019-07-03T18:37:36.771Z] >>> HTTP REQUEST POST https://cloudresourcemanager.googleapis.com/v1/projects/xxxxxxx-dev-eu-west3:testIamPermissions  

[debug] [2019-07-03T18:37:37.606Z] <<< HTTP RESPONSE 200
[debug] [2019-07-03T18:37:37.614Z] [functions] Using node@8 from host.
[debug] [2019-07-03T18:37:37.615Z] >>> HTTP REQUEST GET https://cloudresourcemanager.googleapis.com/v1/projects/xxxxxxx-dev-eu-west3  
 
[debug] [2019-07-03T18:37:38.373Z] <<< HTTP RESPONSE 200
[debug] [2019-07-03T18:37:38.373Z] >>> HTTP REQUEST GET https://mobilesdk-pa.googleapis.com/v1/projects/595396935831:getServerAppConfig  
 
[debug] [2019-07-03T18:37:38.637Z] <<< HTTP RESPONSE 200
[info] ✔  functions: Emulator started at http://localhost:5001
[debug] [2019-07-03T18:37:38.644Z] [functions] Watching "/Users/seb/dev/git/xxxxxxx/xxxxxxx_firebase/Firebase/functions" for Cloud Functions...
[debug] [2019-07-03T18:37:38.935Z] DEBUG: Functions runtime initialized.
[debug] [2019-07-03T18:37:38.935Z] DEBUG: Disabled runtime features: {"functions_config_helper":true,"network_filtering":true,"timeout":true,"memory_limiting":true,"protect_env":true,"admin_stubs":true}
[debug] [2019-07-03T18:37:38.994Z] DEBUG: firebase-admin has been stubbed.
[debug] [2019-07-03T18:37:39.310Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.310Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.310Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.310Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.310Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.310Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.310Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.311Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.311Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.311Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[debug] [2019-07-03T18:37:39.311Z] INFO: Ignoring trigger "xxxxxxx" because the Cloud Firestore emulator is not running.
[info] i  functions: Loaded functions: xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx, xxxxxxx
```

I'm going to try your commands

**Comment by sperochon** [comment]
Here are the results of your commands
```
> firebase serve

=== Serving from '/Users/seb/dev/git/xxxxxxx/xxxxxxx_firebase/Firebase'...

✔  functions: Using node@8 from host.
✔  functions: Emulator started at http://localhost:5000
i  functions: Watching "/Users/seb/dev/git/xxxxxxx/xxxxxxx_firebase/Firebase/functions" for Cloud Functions...
⚠  AssertionError [ERR_ASSERTION]: missing path
    at Module.require (module.js:595:3)
    at require (internal/module.js:11:18)
    at InitializeFirebaseAdminStubs (/Users/seb/.nvm/versions/node/v8.16.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:231:18)
    at /Users/seb/.nvm/versions/node/v8.16.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:461:9
    at Generator.next (<anonymous>)
    at /Users/seb/.nvm/versions/node/v8.16.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/seb/.nvm/versions/node/v8.16.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:3:12)
    at main (/Users/seb/.nvm/versions/node/v8.16.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:431:12)
    at Object.<anonymous> (/Users/seb/.nvm/versions/node/v8.16.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:521:5)
⚠  Your function was killed because it raised an unhandled error.
```

**Comment by samtstern** [comment]
@sperochon the point about the region is really interesting.  So in your code, it breaks 100% if you put `europe-west1` but works if you put `us-east1`?

**Comment by sperochon** [comment]
@samtstern I've investigated deeper and it seems that the region thing was a side effect...

So, I've finally succeed to reproduce the problem in a brand new project ([projectKO.zip](https://github.com/firebase/firebase-tools/files/3364584/projectKO.zip))
I did
`firebase init functions`
to create the project.

I've only replaced the index.js file with this one (contains 15 exactly same functions)

```
const functions = require('firebase-functions');

exports.function1 = functions
  .firestore
  .document('/collection/{item}')
  .onCreate((snap, context) => {
    return Promise.resolve();
  });

exports.function2 = functions
.firestore
.document('/collection/{item}')
.onCreate((snap, context) => {
  return Promise.resolve();
});

...
```

With this project, the command
`firebase functions:shell`
gives
`✔  functions: Emulator started at http://localhost:5001` and get stucks forever



If I comment a function inside the index.js file
```
const functions = require('firebase-functions');
/*
exports.function1 = functions
  .firestore
  .document('/collection/{item}')
  .onCreate((snap, context) => {
    return Promise.resolve();
  });
*/

exports.function2 = functions
.firestore
.document('/collection/{item}')
.onCreate((snap, context) => {
  return Promise.resolve();
});

...
```
then it works:
```
✔  functions: Emulator started at http://localhost:5001
i  functions: Loaded functions: function2, function3, function4, function5, function6, function7, function8, function9, function10, function11, function12, function13, function14, function15
firebase > 
```

**Comment by samtstern** [comment]
@sperochon huh that's really interesting, so in your case it has something to do with the number of functions maybe?

I was not able to observe the same limit on my machine:
```
$ npx firebase functions:shell
✔  functions: Emulator started at http://localhost:5000
i  functions: Loaded functions: helloWorld, function2, function3, function4, function5, function6, function7, function8, function9, function10, function11, function12, function13, function14, function15, function16, function17, function18, function19, function20, function21
```

**Comment by sperochon** [comment]
@samtstern Thanks for your try. I have created a brand new user session on my computer. I've reinstalled nvm, node, firebase-tools, the project but still the same... So it doesn't seem to be a config problem... Did you test with exactly my files project?

**Comment by sperochon** [comment]
And do you know why the port used for me is 5001 and yours 5000?

**Comment by luisatlive** [comment]
I have the same issue; hangs when I have my full 25 functions, if I cut them down to 23 or below it works just fine... this was on firebase-tools 7.0.2. I downgraded to 6.12.0 and it works again now with all functions.

**Comment by sperochon** [comment]
Thanks @luisatlive ! I think you found the guilty!
Downgrading from 'firebase-tools@7.0.2' to 'firebase-tools@7.0.1' works for me on my sample project (projectKO.zip above)! (I'll test it tomorrow on my real project)
My config:
Node v8.16.0
firebase-admin 8.2.0 (latest)
firebase-functions 3.1.0 (latest)

**Comment by luisatlive** [comment]
np; just a heads up once I downgraded I ran into this bug (https://github.com/firebase/firebase-tools/issues/1288) with Firestore triggers, so also had to downgrade firebase-functions to 2.3.1 and now I'm back in a working state.

**Comment by prgwar** [comment]
[2019-07-10T11:48:29.111Z] DEBUG: Functions runtime initialized.
[2019-07-10T11:48:29.112Z] DEBUG: Disabled runtime features: {"functions_config_helper":true,"network_filtering":true,"timeout":true,"memory_limiting":true,"protect_env":true,"admin_stubs":true}
[2019-07-10T11:48:29.226Z] DEBUG: **firebase-admin has been stubbed.**

This is the log I am getting I tried with changing the zone of the function didn't help. I have 55 functions in my project. I referenced the google docs a project can have up to 1000 functions.

**Comment by samtstern** [comment]
@sperochon the reason I have a different port than you is because I am working off of the `master` branch which includes this fix: https://github.com/firebase/firebase-tools/commit/729083ffa6b5ce07c192a5f3171b56bab5f0d3f1

Thank you to everyone for posting your data here!  I'm still trying to reproduce this so I can hunt for the cause.

**Comment by samtstern** [comment]
Also a quick question for everyone who is having this issue: are you able to run `firebase emulators:start --only functions` and see all of your functions emulated or does that hang as well?  I am trying to figure out if this is a `shell`-only issue or a more general Functions emulation issuee.

**Comment by sperochon** [comment]
@samtstern it is the same
```
> firebase emulators:start --only functions --debug
[2019-07-10T16:42:44.968Z] ----------------------------------------------------------------------
[2019-07-10T16:42:44.972Z] Command:       /Users/seb/.nvm/versions/node/v8.16.0/bin/node /Users/seb/.nvm/versions/node/v8.16.0/bin/firebase emulators:start --only functions --debug
[2019-07-10T16:42:44.972Z] CLI Version:   7.0.2
[2019-07-10T16:42:44.972Z] Platform:      darwin
[2019-07-10T16:42:44.972Z] Node Version:  v8.16.0
[2019-07-10T16:42:44.973Z] Time:          Wed Jul 10 2019 18:42:44 GMT+0200 (CEST)
[2019-07-10T16:42:44.973Z] ----------------------------------------------------------------------

[2019-07-10T16:42:44.982Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2019-07-10T16:42:44.982Z] > authorizing via signed-in user
[2019-07-10T16:42:44.984Z] >>> HTTP REQUEST GET https://cloudresourcemanager.googleapis.com/v1/projects/xxxxxx  
 
[2019-07-10T16:42:45.829Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=UTF-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Wed, 10 Jul 2019 16:42:46 GMT, server=ESF, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, server-timing=gfet4t7; dur=664, alt-svc=quic=":443"; ma=2592000; v="46,43,39", accept-ranges=none, transfer-encoding=chunked
i  Starting emulators: ["functions"]
✔  functions: Using node@8 from host.
[2019-07-10T16:42:45.856Z] >>> HTTP REQUEST GET https://mobilesdk-pa.googleapis.com/v1/projects/595396935831:getServerAppConfig  
 
[2019-07-10T16:42:46.115Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=UTF-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Wed, 10 Jul 2019 16:42:46 GMT, server=ESF, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, alt-svc=quic=":443"; ma=2592000; v="46,43,39", accept-ranges=none, transfer-encoding=chunked
✔  functions: Emulator started at http://localhost:5001
i  functions: Watching "/Users/seb/dev/git/xxxxxx/xxxxxx_firebase/Firebase/functions" for Cloud Functions...
[2019-07-10T16:42:46.674Z] Functions runtime initialized.
[2019-07-10T16:42:46.674Z] Disabled runtime features: undefined
[2019-07-10T16:42:46.830Z] Found google-gax at /Users/seb/dev/git/xxxxxx/xxxxxx_firebase/Firebase/functions/node_modules/google-gax/build/src/index.js
[2019-07-10T16:42:46.830Z] Outgoing network have been stubbed.
[2019-07-10T16:42:46.872Z] Checked functions.config()
[2019-07-10T16:42:46.874Z] firebase-admin has been stubbed.
```

**Comment by samtstern** [comment]
@sperochon ok thank you!  That's probably the most useful clue yet.

**Comment by samtstern** [comment]
Wooooot I reproduced this.  Now I can try to fix it without stabbing blindly!

**Comment by samtstern** [comment]
The fix is in #1487

**Comment by sperochon** [comment]
@samtstern Yes! It now works well with firebase-tools 7.1.0  Seems to be a big fix. Thanks and have a good day! :-)
